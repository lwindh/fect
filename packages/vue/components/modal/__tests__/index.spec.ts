import { mount } from '@vue/test-utils'
import Modal from '../modal'

// see : https://github.com/vuejs/vue-test-utils-next/issues/183(test for teleport)

// normal test . in future will support instrcution

const Wrapper = {
  components: {
    [Modal.name]: Modal
  },
  data() {
    return { show: false, mounted: false, overlay: true }
  },
  template: `
  <div class="container">
   <button id="btn" @click="show=true">Button</button>
    <fe-modal v-model:visible="show"
      :overlay="overlay"
      title="Test Modal"
      teleport=".container"
      cancel="Cancel"
      done="Done"
      width="500px"
      v-if="mounted" />
  </div>
  `
}

describe('Modal', () => {
  it('should be support normal props', async () => {
    const wrapper = mount(Wrapper, { attachTo: document.body })
    await wrapper.setData({ show: true })
    await wrapper.setData({ mounted: true })
    await wrapper.setData({ overlay: false })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.title').text()).toBe('Test Modal')
    expect(wrapper.find('.fect-teleport__overlay').exists()).toBe(false)
    expect(wrapper.findAll('.fect-modal__button')[0].text()).toBe('Cancel')
    expect(wrapper.findAll('.fect-modal__button')[1].text()).toBe('Done')
    expect(() => wrapper.unmount).not.toThrow()
    wrapper.unmount()
  })
  it('should be support custom slots', async () => {
    const wrapper = mount(
      {
        components: {
          [Modal.name]: Modal
        },
        data() {
          return { show: false, mounted: false }
        },
        template: `
        <div class="container">
          <fe-modal v-model:visible="show" teleport=".container" v-if="mounted">
            <template #title>Custom Title</template>
            <template #action>
              <h3>Custom Action</h3>
            </template>
          </fe-modal>
        </div>
      `
      },
      { attachTo: document.body }
    )

    await wrapper.setData({ show: true })
    await wrapper.setData({ mounted: true })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.fect-modal__title').text()).toBe('Custom Title')
    expect(wrapper.find('h3').text()).toBe('Custom Action')
    wrapper.unmount()
  })

  it('should be close modal through props visible value is false', async () => {
    const wrapper = mount(Wrapper, { attachTo: document.body })
    const btn = wrapper.find('#btn')
    await wrapper.setData({ mounted: true })
    await btn.trigger('click')
    expect(wrapper.find('.fect-modal__root').isVisible()).toBe(true)
    const el = wrapper.findAll('.fect-modal__button')
    await el[0].trigger('click')
    expect(wrapper.vm.show).toBe(false)
    expect(wrapper.find('.fect-modal__root').isVisible()).toBe(false)
    await wrapper.setData({ show: true })
    await wrapper.setData({ mounted: true })
    expect(wrapper.vm.show).toBe(true)
    await wrapper.find('.fect-modal__root').trigger('click')
    expect(wrapper.vm.show).toBe(false)
    await wrapper.setData({ show: true })
    await wrapper.setData({ mounted: true })
    await wrapper.find('.fect-modal__wrapper').trigger('click')
    expect(wrapper.vm.show).toBe(true)
    wrapper.unmount()
  })
})
