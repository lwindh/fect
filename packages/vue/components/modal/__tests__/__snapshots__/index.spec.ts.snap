// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should be support custom slots 1`] = `
<div class="container">
  <!--teleport start-->
  <!--teleport end-->
  <transition-stub>
    <div class="fect-teleport__overlay"></div>
  </transition-stub>
  <transition-stub>
    <div class="fect-modal__root" role="popup">
      <div role="dialog" tabindex="-1" class="fect-modal__wrapper" style="width: 400px;">
        <div class="fect-modal__title">Custom Title</div>
        <div class="fect-modal__content">
          <!---->
        </div>
        <h3>Custom Action</h3>
      </div>
    </div>
  </transition-stub>
</div>
`;

exports[`Modal should be support normal props 1`] = `
<div class="container">
  <!--teleport start-->
  <!--teleport end-->
  <!---->
  <transition-stub>
    <div class="fect-modal__root" role="popup">
      <div role="dialog" tabindex="-1" class="fect-modal__wrapper" style="width: 500px;">
        <div class="fect-modal__title">
          <h2 class="title">Test Modal</h2>
        </div>
        <div class="fect-modal__content">
          <!---->
        </div>
        <footer class="fect-modal__action"><button class="fect-button fect-button--default fect-button--medium fect-modal__button" style="--button-hover-bg: var(--primary-background); --button-hover-border: var(--primary-foreground); --button-hover-color: var(--primary-foreground);">
            <!---->
            <!---->
            <!---->Cancel
          </button><button class="fect-button fect-button--default fect-button--medium fect-modal__button" style="--button-hover-bg: var(--primary-background); --button-hover-border: var(--primary-foreground); --button-hover-color: var(--primary-foreground);">
            <!---->
            <!---->
            <!---->Done
          </button></footer>
      </div>
    </div>
  </transition-stub>
</div>
`;
